### PR TITLE
Allow wall dropoff

### PR DESCRIPTION
**Pull request by kepler471**
Houston, we have lift off (from walls).

Fixed a load of problems with this pull.

Have a new, 3rd RayCast2D, whch gives much more flexibilty in determining the state of the player. No longer flips between Wall and Air (running 1 physics_process per flip) at any transition or state that I have tested!!

Have removed Air -> Ledge transitions which has done mad improvements for the maneuvering around single 32px block height increases.

### CODE REVIEW

File: Moonshot/player/States/air.gd [hunk 1]
```diff
@@ -32,15 +32,15 @@ func physics_process(delta: float) -> void:
...
 	if owner.is_on_floor():
 		var target_state := "Move/Idle" if _parent.get_move_direction().x == 0 else "Move/Run"
 		_state_machine.transition_to(target_state)
 
-	elif owner.ledge_wall_detector.is_against_ledge():
-		_state_machine.transition_to("Ledge", {move_state = _parent})
-
-	if owner.is_on_wall():
+	# Wall hanging dropoff
+	elif owner.is_on_wall() and ld.is_against_wall() and not ld.is_hanging():
```
Comment (kepler471) [inline Moonshot/player/States/air.gd:43]: This is the juicy dropoff logic

File: Moonshot/player/States/wall.gd [hunk 2]
```diff
@@ -25,14 +25,19 @@ func physics_process(delta: float) -> void:
...
 	if owner.is_on_floor():
 		_state_machine.transition_to("Move/Idle")
 
-	var is_moving_away_from_wall := sign(_parent.get_move_direction().x) == sign(_wall_normal)
-	if is_moving_away_from_wall:
-		_state_machine.transition_to("Move/Air", {"velocity":_velocity})
+	elif is_moving_away_from_wall:
+		_state_machine.transition_to("Move/Air", {"velocity": _velocity})
+
+	elif ld.is_hanging():
```
Comment (kepler471) [inline Moonshot/player/States/wall.gd:37]: dropoff point

File: Moonshot/player/LedgeWallDetector.gd [hunk 3]
```diff
@@ -23,20 +25,32 @@ func _unhandled_input(event: InputEvent) -> void:
```
Comment (Jamesbarford) [inline Moonshot/player/LedgeWallDetector.gd:26]: Maybe for consistency with these methods call the ray cast methods in the order they are attached to the player:

```
ray_top
ray_middle
ray_bottom
```

Just incase we have to debug this code again then ordering of the methods is one less thing to keep track of

**Review by Jamesbarford — CHANGES_REQUESTED** [review]

**Review by whitew1994WW — APPROVED** [review]
Looking good boi

**Comment by Jamesbarford** [comment]
Feel wise though 👌, so nice.

File: Moonshot/player/LedgeWallDetector.gd (hunk 3 above)
Comment (kepler471) [reply Moonshot/player/LedgeWallDetector.gd:26]: makes sense

**Pull request merged**